src/main/scala/is/hail/io/index/IndexWriter.scala

```diff
@@ -0,0 +1,164 @@
+  private val internalEncoder = codecSpec.buildEncoder(InternalNodeBuilder.typ(keyType, annotationType))(trackedOS)
+
+  private def calcDepth: Int =
+    math.max(1, (math.log10(elementIdx) / math.log10(branchingFactor)).ceil.toInt) // max necessary for array of length 1 becomes depth = 0
+
+  // last internal node builder is the root (not written): internal node length - root + leaf node == internalNodes.length
+  private def height: Int = internalNodeBuilders.length
+
+  private def writeInternalNode(node: InternalNodeBuilder): IndexNodeInfo = {
+    val indexFileOffset = trackedOS.bytesWritten
+
+    assert(node.size > 0)
+    val child = node.getChild(0)
+    val firstIndex = node.firstIdx
+    val firstKey = child.firstKey
+    val firstRecordOffset = child.firstRecordOffset
+    val firstAnnotation = child.firstAnnotation
+
+    internalEncoder.writeByte(1)
+
+    val regionOffset = node.write(rvb)
+    internalEncoder.writeRegionValue(region, regionOffset)
+    internalEncoder.flush()
+
+    region.clear()
+    node.clear(elementIdx)
+
+    IndexNodeInfo(indexFileOffset, firstIndex, firstKey, firstRecordOffset, firstAnnotation)
+  }
+
+  private def writeLeafNode(): IndexNodeInfo = {
+    val indexFileOffset = trackedOS.bytesWritten
+
+    assert(leafNodeBuilder.size > 0)
+    val child = leafNodeBuilder.getChild(0)
+    val firstIndex = leafNodeBuilder.firstIdx
+    val firstKey = child.key
+    val firstRecordOffset = child.recordOffset
+    val firstAnnotation = child.annotation
+
+    leafEncoder.writeByte(0)
+
+    val regionOffset = leafNodeBuilder.write(rvb)
+    leafEncoder.writeRegionValue(region, regionOffset)
+    leafEncoder.flush()
+
+    region.clear()
+    leafNodeBuilder.clear(elementIdx)
+
+    IndexNodeInfo(indexFileOffset, firstIndex, firstKey, firstRecordOffset, firstAnnotation)
+  }
+
+  private def write(flush: Boolean = false) {
+    val nInternalNodes = math.max(calcDepth - 1, 1)
+
+    def writeInternalNodes(level: Int) {
+      if (level < nInternalNodes) {
+        val node = internalNodeBuilders(level)
+
+        if (node.size == branchingFactor || (flush && node.size > 0)) {
+          val info = writeInternalNode(node)
+          if (level + 1 == internalNodeBuilders.size)
+            internalNodeBuilders += new InternalNodeBuilder(keyType, annotationType, info.firstIndex)
+          internalNodeBuilders(level + 1) += info
+          writeInternalNodes(level + 1)
+        } else if (flush)
+          writeInternalNodes(level + 1)
+      }
+    }
+
+    if (leafNodeBuilder.size == branchingFactor || (flush && leafNodeBuilder.size > 0)) {
+      val info = writeLeafNode()
+      if (internalNodeBuilders.isEmpty)
+        internalNodeBuilders += new InternalNodeBuilder(keyType, annotationType, info.firstIndex)
+      internalNodeBuilders(0) += info
+      writeInternalNodes(0)
+    } else if (flush && internalNodeBuilders.size > 0)
```

Comment:
If you require an internal block, you can eliminate the second half of this condition.